python/pyarrow/gandiva.pyx

```diff
@@ -0,0 +1,184 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# cython: profile=False
+# distutils: language = c++
+# cython: embedsignature = True
+
+from libcpp cimport bool as c_bool, nullptr
+from libcpp.memory cimport shared_ptr, unique_ptr, make_shared
+from libcpp.string cimport string as c_string
+from libcpp.vector cimport vector as c_vector
+from libc.stdint cimport int64_t, uint8_t, uintptr_t
+
+from pyarrow.includes.libarrow cimport *
+from pyarrow.compat import frombytes
+from pyarrow.lib cimport (check_status, pyarrow_wrap_array)
+
+from pyarrow.includes.libgandiva cimport (CCondition, CExpression,
+                                          CNode, CProjector, CFilter,
+                                          CSelectionVector,
+                                          TreeExprBuilder_MakeExpression,
+                                          TreeExprBuilder_MakeFunction,
+                                          TreeExprBuilder_MakeLiteral,
+                                          TreeExprBuilder_MakeField,
+                                          TreeExprBuilder_MakeIf,
+                                          TreeExprBuilder_MakeCondition,
+                                          SelectionVector_MakeInt32,
+                                          Projector_Make,
+                                          Filter_Make)
+
+from pyarrow.lib cimport (Array, DataType, Field, MemoryPool,
+                          RecordBatch, Schema)
+
+cdef class Node:
```

Comment:
We also need to prevent instantiating python objects with uninitialized cdef properties. It requires to use both cython's fast instantiation and raising an exception from __init__ like here.

Reply:
Thanks, fixed it!